src/Futhark/Internalise.hs

```diff
@@ -880,6 +904,57 @@ internaliseExp _ e@E.ProjectSection{} =
 internaliseExp _ e@E.IndexSection{} =
   fail $ "internaliseExp: Unexpected index section at " ++ locStr (srclocOf e)
 
+boolOp :: VName -> E.Exp -> E.Exp -> E.Exp
+boolOp op l r = E.BinOp (qualName op) (Info $ vacuousShapeAnnotations ft)
+                 (l, sType l) (r, sType r) (Info (E.Prim E.Bool)) noLoc
+  where sType e = Info $ toStruct $ vacuousShapeAnnotations $ E.typeOf e
+        arrow   = Arrow S.empty Nothing
+        ft      = E.typeOf l `arrow` E.typeOf r `arrow` E.Prim E.Bool
```

Comment:
This entire thing is pretty odd.  You construct a pretty baroque mechanism (also `vnM` below) to construct `&&` and `==` expressions.  Why not just generate `If` expressions for the former?  They will be internalised the same way.  I'd suggest just defining two functions `andExp, eqExp :: E.Exp -> E.Exp -> E.Exp`.

Reply:
`andExp` was easy enough to do, but with `eqExp` I'm still not really sure how to handle the `VName` stuff (in regards to `vnM`). The tag doesn't actually matter here (as long as it's less than `maxIntrinsicTag`), so I could just put `(-1)` and call it a day, but that seems a bit jank to me. What do you suggest?

Reply:
The nicest way would be to generate core code, but I guess that's a bit awkward.  Using `-1` will suffice for now.